### PR TITLE
lint: Use consistent out-of-tree build for python and test_runner

### LINKED ISSUE #30496: Spurious markdown linter failure
The linter job failed for https://github.com/bitcoin/bitcoin/pull/30440

```
One or more markdown links are broken.

Relative links are preferred (but not required) as jumping to file works natively within Emacs.

Markdown link errors found:
[Err ] ./pyenv/README.md (12, 1) => /terminal_output.png - Target filename not found.

                
^---- ⚠️ Failure generated from lint check 'markdown'!
Check that markdown links resolve
```

The PR does not touch any markdown files. 

https://cirrus-ci.com/task/6591959924473856

### PR DESCRIPTION
**Pull request by maflcko**
Fixes https://github.com/bitcoin/bitcoin/issues/30496

Seems odd to sometimes do an out-of-tree build (via `./ci/lint_imagefile`, see `test/lint/README.md`) and sometimes not (via Cirrus CI, see `./ci/lint_run_all.sh`).

Fix it by doing an out-of-tree build consistently in the same location.

Also, fix `$CI_RETRY_EXE`, while touching this.

### CODE REVIEW

**Comment by paplorinc** [comment]
Thanks for fixing this so quickly!
I've added https://github.com/bitcoin/bitcoin/pull/30500 to fix the unrelated warnings that made it more difficult to see what the underlying problem was here.

**Review by Mahmoud198425 — APPROVED** [review]

File: ci/lint/04_install.sh [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 #
-# Copyright (c) 2018-2022 The Bitcoin Core developers
+# Copyright (c) 2018-present The Bitcoin Core developers
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 export LC_ALL=C
 
-export PATH=$PWD/ci/retry:$PATH
+export CI_RETRY_EXE="/ci_retry --"
```
Comment (paplorinc) [inline ci/lint/04_install.sh:9]: Now that we're setting this explicitly, does the prefixing of `CI_RETRY_EXE` still make sense in every case?
If it does, would it make sense to use it for e.g. `${CI_RETRY_EXE} curl` (like in https://github.com/bitcoin/bitcoin/blob/master/ci/test/01_base_install.sh#L88) or `${CI_RETRY_EXE} cargo build` commands as well?

File: ci/lint/04_install.sh [hunk 2]
```diff
@@ -62,3 +64,5 @@ MLC_VERSION=v0.18.0
 MLC_BIN=mlc-x86_64-linux
 curl -sL "https://github.com/becheran/mlc/releases/download/${MLC_VERSION}/${MLC_BIN}" -o "/usr/bin/mlc"
 chmod +x /usr/bin/mlc
+
+popd || exit
```
Comment (paplorinc) [inline ci/lint/04_install.sh:68]: can `popd` fail if `pushd` succeeded?

File: ci/lint/04_install.sh [hunk 3]
```diff
@@ -62,3 +64,5 @@ MLC_VERSION=v0.18.0
 MLC_BIN=mlc-x86_64-linux
 curl -sL "https://github.com/becheran/mlc/releases/download/${MLC_VERSION}/${MLC_BIN}" -o "/usr/bin/mlc"
```
Comment (paplorinc) [inline ci/lint/04_install.sh:65]: Unrelated, but shouldn't git clone contain `--depth=1` throughout this file?

**Review by l0rinc — APPROVED** [review]
utACK fa8d73e86e1c11cdfe8154ab84edc1948283454b

File: ci/lint/04_install.sh (hunk 2 above)
Comment (paplorinc) [reply ci/lint/04_install.sh:68]: k, thanks

**Comment by willcl-ark** [comment]
utACK fa8d73e86e1c11cdfe8154ab84edc1948283454b

Thanks for addressing @maflcko, this makes things much more sanitary.

That said, our `mlc` lint is _supposed_ to exclude all files/dirs which are not tracked by git, by calling `git ls-files --ignored --others --exclude-standard`.

It appears that the `--ignored` flag causes some directories/files to not appear in this listing, leading to the failure:

```bash
will@ubuntu in ~/src/bitcoin on  2407-lint-fixes [$?] : 🐍 (bitcoin)
$ git ls-files --ignored --others --exclude-standard | rg pyenv

will@ubuntu in ~/src/bitcoin on  2407-lint-fixes [$?] : 🐍 (bitcoin)
✗ git ls-files --others --exclude-standard | rg pyenv
.pyenv/readme.md
```

I'll try an upstream fix so that we can avoid a future recurrence (with another file/dir)

**Pull request merged**